### PR TITLE
Experiment with github actions for CI

### PR DESCRIPTION
**Pull request by Osspial**

### CODE REVIEW

**Comment by Osspial** [comment]
I quite like how progress is going on this. Based on my experience so far, I think we should absolutely aim to switch from Travis and Appveyor to GitHub actions once we can work the kinks out. As a plus, we'll be able to more easily run tests on iOS and Android, since GitHub's images seem to come with android and iOS simulators installed.

That said, the new system is running into some failures the old system didn't, since it's testing stuff more comprehensively. As a quick rundown:

- iOS tests are failing because of #1310. I don't know if we'll have to do additional special-casing to get the iOS simulator running.
- `web-sys` (*not* `stdweb`) is failing tests with the following error:
    ```
    Error loading Rust wasm module 'winit_4a8d91107b1f9f84': <TypeError>
    error: unhandled exception thrown
    error:     TypeError: WebAssembly.instantiate(): Import #0 module="__wbindgen_placeholder__" error: module is not an object or function
        at WebAssembly.instantiate (http://localhost:33217/__cargo-web__/test_runner.js:319:36)
        at http://localhost:33217/js/app.js:37:65
    error: source: http://localhost:33217/__cargo-web__/test_runner.js:318:35
    ```

Those are the only real outstanding CI failures, though. We'll also need to get auto-publish working in github actions, but I expect that'll be cleaner to do through the actions system than through Travis.

**Comment by goddessfreya** [comment]
@Osspial Was hoping to get auto publishing working on glutin and winit, but I need access to the secrets tab in the settings (currently don't have access to the settings at all)

**Comment by goddessfreya** [comment]
@simlay Can you resolve this new iOS build failure?

https://github.com/rust-windowing/winit/pull/1309/checks?check_run_id=360305405#step:8:195

**Comment by Osspial** [comment]
I'll take care of getting auto publishing working once the rest of the CI works

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+name: Rust
+
+on: [push]
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Add `pull_request` else it won't work on pull requests.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        rust-version: stable
+        components: rustfmt
+    - name: Check Formatting
+      run: cargo +stable fmt --all -- --check
+
+  test:
+    strategy:
+      fail-fast: false
+      matrix:
+        rust_version: [stable]
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Test nightly too?

**Comment by simlay** [comment]
> @simlay Can you resolve this new iOS build failure?
> 
> https://github.com/rust-windowing/winit/pull/1309/checks?check_run_id=360305405#step:8:195

It seems to be failing to build the tests in the multithreaded issue fixed in #1310, I cherry-picked the commit and it worked until it actually needed to run the test... I don't fully know how `cargo  test --verbose --target aarch64-apple-ios --features ','` builds the binary for the tests but the binary it builds is for `aarch64-apple-ios` which I think can only work in the iOS simulator or an actual device.

The only CI setup I know of for doing iOS rust testing stuff is [how this guy did it](http://sasheldon.com/blog/2016/02/22/testing-rust-on-ios-with-travis/).

For the iOS targets, we should probably just do `--no-run` for the tests.

**Comment by goddessfreya** [comment]
> > [](https://github.com/rust-windowing/winit/pull/1309/checks?check_run_id=360305405#step:8:195)
> 
> It seems to be failing to build the tests in the multithreaded issue fixed in #1310, I cherry-picked the commit and it worked until it actually needed to run the test...

Ah, I see. I guess @Osspial needs to just push another commit so it builds against a new master.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Osspial) [reply .github/workflows/ci.yml]: I'll test the other channels once this is ready to merge. I'd cut it back to just stable to speed up the edit/debug cycle.

**Comment by Osspial** [comment]
There seem to be some duplicated tests running on non-Windows platforms? I think the matrix might be a bit wonky

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        platform:
+          - { os: windows-latest, target: x86_64-pc-windows-msvc,   features: ',' }
+          - { os: windows-latest, target: i686-pc-windows-msvc,     features: ',' }
+          - { os: windows-latest, target: x86_64-pc-windows-gnu,    features: ',', host: -x86_64-pc-windows-gnu }
+          - { os: windows-latest, target: i686-pc-windows-gnu,      features: ',', host: -i686-pc-windows-gnu }
+          - { os: ubuntu-latest,  target: i686-unknown-linux-gnu,   features: ',' }
+          - { os: ubuntu-latest,  target: i686-unknown-linux-gnu,   features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Well, you are duplicating them in the matrix. This one is the same as the one above.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+          - { os: ubuntu-latest,  target: x86_64-unknown-linux-gnu, features: ',' }
+          - { os: ubuntu-latest,  target: x86_64-unknown-linux-gnu, features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: And this one is the same as the one above it.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+          - { os: macos-latest,   target: x86_64-apple-darwin,      features: ',' }
+          - { os: macos-latest,   target: x86_64-apple-darwin,      features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Ditto.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          - { os: macos-latest,   target: x86_64-apple-ios,         features: ',' }
+          - { os: macos-latest,   target: x86_64-apple-ios,         features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Ditto

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+          - { os: macos-latest,   target: armv7-apple-ios,          features: ',' }
+          - { os: macos-latest,   target: armv7-apple-ios,          features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Ditto

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+          - { os: macos-latest,   target: aarch64-apple-ios,        features: ',' }
+          - { os: macos-latest,   target: aarch64-apple-ios,        features: ',' }
```
Comment (goddessfreya) [inline .github/workflows/ci.yml]: Ditto

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Osspial) [reply .github/workflows/ci.yml]: 🤦‍♀️

**Comment by Osspial** [comment]
@ryanisaacg could you look into why the `web-sys` tests are failing to run? See [here](https://github.com/rust-windowing/winit/pull/1309/checks?check_run_id=373451657) for an example.

**Comment by ryanisaacg** [comment]
@Osspial The error isn't super clear to me, but the problem is almost certainly using `cargo web` to test `web-sys` (`cargo web` works with just `stdweb`). We could use [`wasm-pack`'s test runner](https://rustwasm.github.io/docs/wasm-pack/commands/test.html) but that might be overkill, because we currently have no browser tests.

**Comment by Osspial** [comment]
@ryanisaacg ah, fair enough. Most of the utility from the current tests comes from compiling them anyway, so I've disabled actually running the tests for wasm platforms.

**Pull request merged**